### PR TITLE
Implement event filters and clean up code a little

### PR DESCRIPTION
**Pull request by ryepdx**

### CODE REVIEW

**Comment by tcoulter** [comment]
Wow, this looks like great work. I'll give it a deeper look soon.

File: testrpc/testrpc.py [hunk 1]
```diff
@@ -83,13 +154,94 @@ def format_block_number(block_number):
...
+            "tx_idx": len(block.get_transactions()), "txhash": evm.last_tx.hash,
+            "pending": False, "block": block
+        })
+        self.log_idx += 1
+
+def mine(evm):
+    if evm.block.number == 0:
+        evm.block.log_listeners.append(LogListener())
+
+    listener = evm.block.log_listeners.pop(0)
```
Comment (chetan) [inline testrpc/testrpc.py]: Why do you pop out the listener before mining? Doesn't that stop the log events from being collected?

Comment (ryepdx) [inline testrpc/testrpc.py]: No, but I can see why you'd be concerned. The "mine" function on the block doesn't actually log anything. All it does is instantiate a new "block" object to collect transactions in. The events are logged when the transactions come in, not when they're mined. That's not how it works on the real blockchain, of course, so I'll go ahead and change this bit to be less confusing. I ultimately don't really have a good reason for popping the listeners off the block, apart from trying to save some memory as the blockchain grows, which is totally a premature optimization. And it looks like PyEthereum carries forward the log listeners upon mining, so I should be able to remove the `append` line as well.

Comment (chetan) [inline testrpc/testrpc.py]: Yeah, I actually tried commenting it out and it works fine either way.

**Comment by ryepdx** [comment]
Removed both lines, appears to be working fine. Looks like there are merge conflicts now, though.

**Comment by ryepdx** [comment]
Merged in your changes to the master branch and resolved the merge conflicts. Anything else I need to do before this is ready to be merged in?

**Comment by tcoulter** [comment]
Sorry, this is my fault. Since I last commented I've moved across the country and this got lost. Will take a look and comment/merge today.

**Comment by tcoulter** [comment]
@ryepdx Everything is looking great, and thanks for your work. Question, since I'm not a python expert: You made the following commit:

```
f0fa961fa4	Use Python 3 compatible print function
```

This print function is still compatible with Python 2, correct, and doesn't require Python 3 in any way? It doesn't matter so much to me personally, but I know other users who only have Python 2.x have complained that a Python 3 module was used/required (still figuring this issue out). Thanks! 

Will merge once I receive confirmation from you.

**Comment by tcoulter** [comment]
Also, one last thing: Can you update the readme with your changes? For instance, `eth_getFilterChanges` still has "only supports block filters" next to it. You should also add any functions you've added to that list, if any.

Thanks again!

**Comment by ryepdx** [comment]
Yeah, the print function is still compatible with Python 2. There's a way to write Python code so it's compatible with both Python 2 and Python 3. Pyethereum is written this way and is thus cross-release compatible.

**Comment by tcoulter** [comment]
Alright @ryepdx, great work. I'm going to go ahead and merge. Will give this a test in the next couple days then release a new version to pypi within the week.

**Comment by tcoulter** [comment]
Also @ryepdx, if you're interested: https://consensys.net/open-positions/?bzid=e3232ab8e275

**Pull request merged**